framework/i18n/CLocale.php

```diff
@@ -113,7 +113,7 @@ protected function __construct($id)
 	 */
 	public static function getCanonicalID($id)
 	{
-		return strtolower(str_replace('-','_',$id));
+		return strtolower(str_replace('-','_',(string)$id));
```

Comment:
You should not pass `null` to this method according to phpdoc. In what scenario you will get deprecation error here?

Reply:
> You should not pass null to this method according to phpdoc

I agree, it would probably be better to change the test passing `null` to it in the first place. We could either:
**A:** Cast the provided data to a string, for example:
https://github.com/yiisoft/yii/blob/1c6f6fe221404de9afeb32059e6671676237482c/tests/framework/i18n/CLocaleTest.php#L126
Would become:
```php
 $this->assertEquals(mb_strtolower((string)$assertion),mb_strtolower((string)$locale->getLocaleDisplayName((string)$methodLocale))); 
 ```
Note that the first two casts are required as it is possible for [`getLocaleDisplayName`](https://github.com/yiisoft/yii/blob/1c6f6fe221404de9afeb32059e6671676237482c/framework/i18n/CLocale.php#L410-L418) to return `null` into `mb_strtolower`, which has been deprecated.

Or **B:** not pass `null` into `getLocaleDisplayName` at all, as it isn't allowed according to the phpdoc. This would mean that the dataProviders should be changed, for example:
https://github.com/yiisoft/yii/blob/1c6f6fe221404de9afeb32059e6671676237482c/tests/framework/i18n/CLocaleTest.php#L107-L108
Could be changed to something like:
```php
 array('es_MX','',null),  // Empty string, essentially the same as (string)null
 array('es_ES','',null),  // Or any invalid ID causing getLocaleDisplayName to return null
```

@rob006 what are your thoughts on this? I think option **B** might be somewhat clearer for anyone reading the tests

Reply:
I would remove these tests completely, they look invalid to me (they do not respect types from phpdoc). But maybe I'm missing something, so I would wait for second opinion on this.

Reply:
If only tests code calls this function with `null` and never framework code, I would remove the tests. Otherwise, I would vote to keep current behavior and preventing possible regressions by casting to string.

Reply:
The function defaults `$id` to null:
https://github.com/yiisoft/yii/blob/1c6f6fe221404de9afeb32059e6671676237482c/framework/i18n/CLocale.php#L418

Which passes `$id` on to:
https://github.com/yiisoft/yii/blob/1c6f6fe221404de9afeb32059e6671676237482c/framework/i18n/CLocale.php#L108-L114
Which doesn't allow for `$id` to be null.

I think it might be a good idea to cast to string to prevent possible regressions, because `getLocaleDisplayName` is public and allows `$id` to be `null`. Bven though no framework functions call `getLocaleDisplayName` without parameters or with `null` . I'm not quite sure where the best place to cast would be:
1. In the test, although this would not cover `getLocaleDisplayName` without any parameters;
2. In `getLocaleDisplayName`, while passing it to `getCanonicalID`;
3. Or in `getCanonicalID` as it currently is in my fork.

Reply:
Since a public framework function allows $null and passes it on, I consider this a call from framework code. So I suggest to keep the current behavior by casting to string as you currently have it.

Reply:
Accepting `null` in `getLocaleDisplayName()` looks nonsensical IMO - canonicalizing ID is useful only for fetching language name and it will always return `null` in that case (there is no name for "null-language"), so there is no point of calling `getCanonicalID()` at all. IMO we should not propagate such weirdness from `getLocaleDisplayName()` to `getCanonicalID()` and deal with `null` locally :stuck_out_tongue:.

Reply:
@rob006 I agree it's not nice from a design point of view and an inconcistency, trivial as it may be. @wtommyw I'm fine with either solution and leave the final decision up to you. :)